### PR TITLE
#536: Add AddressSanitizer flags to CI tests

### PR DESCRIPTION
**Pull request by PhilMiller**
Additions and changes to the codebase should not be allowed to introduce new errors detected by runtime dynamic analysis by AddressSanitizer.

## Additions

- Address Sanitizer is now activated on all tests run in CI
- Geopackage units tests are run in CI (they were previously missing)

## Changes

- Fixed several errors detected by ASan in GeoPackage code and its tests
- Suppress ASan report from zombie GIUH code

## Testing

This expands what CI covers

## Notes

- Leak detection in ASan is disabled across the board, because there are a number of small leaks that bite every test, which then cause it to abort rather than reporting the test results

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [x] Linux
- [x] macOS

### CODE REVIEW

**Comment by PhilMiller** [comment]
Moving this to completion is blocked on revisiting #539, and possibly also on #567

**Comment by PhilMiller** [comment]
Turns out that `-fsanitize=address` also triggers #567 just like optimization does, so I've disabled sanitizer instrumentation for the affected file `wkb.cpp` as well.

Thus, this PR is no longer blocked on that issue.

**Comment by PhilMiller** [comment]
Seeing some ASan failures in GeoPackage tests, which I didn't cover when I started looking into this, since that code was added more recently.

**Comment by PhilMiller** [comment]
Here's the remaining failure I see from GeoPackage code:

```
[ RUN      ] GeoPackage_Test.geopackage_read_test
Reading 2 features in layer test
=================================================================
==30446==ERROR: AddressSanitizer: container-overflow on address 0x6060000cc4a0 at pc 0x000106df7778 bp 0x00016b225290 sp 0x00016b224a40
READ of size 16 at 0x6060000cc4a0 thread T0
    #0 0x106df7774 in __asan_memcpy+0x37c (libclang_rt.asan_osx_dynamic.dylib:arm64+0x4f774) (BuildId: c7b956d5d2e13c75aade59781ac6571732000000200000000100000000000b00)
    #1 0x104cd8fe0 in void std::__1::vector<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > >::__construct_at_end<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >*, 0>(boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >*, boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >*, unsigned long) vector:1029
    #2 0x104cd8de4 in std::__1::vector<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > >::vector(std::__1::vector<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > > const&) vector:1194
    #3 0x10518fadc in std::__1::__shared_ptr_emplace<geojson::LineStringFeature, std::__1::allocator<geojson::LineStringFeature> >::__shared_ptr_emplace[abi:v160006]<boost::geometry::model::linestring<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::vector, std::__1::allocator>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, geojson::JSONProperty, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, geojson::JSONProperty> > >&, std::__1::vector<double, std::__1::allocator<double> >&>(std::__1::allocator<geojson::LineStringFeature>, boost::geometry::model::linestring<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::vector, std::__1::allocator>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, geojson::JSONProperty, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, geojson::JSONProperty> > >&, std::__1::vector<double, std::__1::allocator<double> >&) shared_ptr.h:287
    #4 0x10518e3d0 in geopackage::build_feature(geopackage::sqlite_iter const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) feature.cpp:36
    #5 0x1051942a0 in geopackage::read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) read.cpp:94
    #6 0x104dbe0cc in GeoPackage_Test_geopackage_read_test_Test::TestBody() GeoPackage_Test.cpp:41
    #7 0x104f5d5d0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) gtest.cc:2635
    #8 0x104f5d304 in testing::Test::Run() gtest.cc:2674
    #9 0x104f60568 in testing::TestInfo::Run() gtest.cc:2853
    #10 0x104f62648 in testing::TestSuite::Run() gtest.cc:3012
    #11 0x104f99a74 in testing::internal::UnitTestImpl::RunAllTests() gtest.cc:5870
    #12 0x104f98b00 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) gtest.cc:2635
    #13 0x104f98890 in testing::UnitTest::Run() gtest.cc:5444
    #14 0x104fc1fc4 in main gtest_main.cc:51
    #15 0x10620d088 in start+0x204 (dyld:arm64+0x5088) (BuildId: a2ee361189123e1dbf8bfed54862d4c932000000200000000100000000060c00)
    #16 0x471f7ffffffffffc  (<unknown module>)

0x6060000cc4a0 is located 0 bytes inside of 64-byte region [0x6060000cc4a0,0x6060000cc4e0)
allocated by thread T0 here:
    #0 0x106e07f80 in wrap__Znwm+0x6c (libclang_rt.asan_osx_dynamic.dylib:arm64+0x5ff80) (BuildId: c7b956d5d2e13c75aade59781ac6571732000000200000000100000000000b00)
    #1 0x1053dea14 in void* std::__1::__libcpp_operator_new[abi:v160006]<unsigned long>(unsigned long) new:266
    #2 0x1053de9f0 in std::__1::__libcpp_allocate[abi:v160006](unsigned long, unsigned long) new:292
    #3 0x1053fafd0 in std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >::allocate[abi:v160006](unsigned long) allocator.h:115
    #4 0x1053faeac in std::__1::__allocation_result<std::__1::allocator_traits<std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > >::pointer> std::__1::__allocate_at_least[abi:v160006]<std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > >(std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >&, unsigned long) allocate_at_least.h:55
    #5 0x1053fadec in std::__1::__split_buffer<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >&) __split_buffer:323
    #6 0x1053faca8 in std::__1::__split_buffer<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > >&) __split_buffer:319
    #7 0x1053dafd0 in std::__1::vector<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> >, std::__1::allocator<boost::geometry::model::point<double, 2ul, boost::geometry::cs::geographic<boost::geometry::degree> > > >::reserve(unsigned long) vector:1503
    #8 0x1053dae0c in geopackage::wkb::wgs84::operator()(boost::geometry::model::linestring<boost::geometry::model::point<double, 2ul, boost::geometry::cs::cartesian>, std::__1::vector, std::__1::allocator>&) wkb.cpp:259
    #9 0x105098888 in geopackage::build_geometry(geopackage::sqlite_iter const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<double, std::__1::allocator<double> >&) geometry.cpp:80
    #10 0x10518e0f0 in geopackage::build_feature(geopackage::sqlite_iter const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) feature.cpp:21
    #11 0x1051942a0 in geopackage::read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) read.cpp:94
    #12 0x104dbe0cc in GeoPackage_Test_geopackage_read_test_Test::TestBody() GeoPackage_Test.cpp:41
    #13 0x104f5d5d0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) gtest.cc:2635
    #14 0x104f5d304 in testing::Test::Run() gtest.cc:2674
    #15 0x104f60568 in testing::TestInfo::Run() gtest.cc:2853
    #16 0x104f62648 in testing::TestSuite::Run() gtest.cc:3012
    #17 0x104f99a74 in testing::internal::UnitTestImpl::RunAllTests() gtest.cc:5870
    #18 0x104f98b00 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) gtest.cc:2635
    #19 0x104f98890 in testing::UnitTest::Run() gtest.cc:5444
    #20 0x104fc1fc4 in main gtest_main.cc:51
    #21 0x10620d088 in start+0x204 (dyld:arm64+0x5088) (BuildId: a2ee361189123e1dbf8bfed54862d4c932000000200000000100000000060c00)
    #22 0x471f7ffffffffffc  (<unknown module>)

HINT: if you don't care about these errors you may set ASAN_OPTIONS=detect_container_overflow=0.
If you suspect a false positive see also: https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow.
SUMMARY: AddressSanitizer: container-overflow (libclang_rt.asan_osx_dynamic.dylib:arm64+0x4f774) (BuildId: c7b956d5d2e13c75aade59781ac6571732000000200000000100000000000b00) in __asan_memcpy+0x37c
Shadow bytes around the buggy address:
  0x6060000cc200: 00 00 00 00 00 00 00 fa fa fa fa fa 00 00 00 00
  0x6060000cc280: 00 00 00 fa fa fa fa fa 00 00 00 00 00 00 00 fa
  0x6060000cc300: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x6060000cc380: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x6060000cc400: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
=>0x6060000cc480: fa fa fa fa[fc]fc fc fc fc fc fc fc fa fa fa fa
  0x6060000cc500: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x6060000cc580: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x6060000cc600: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x6060000cc680: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x6060000cc700: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==30446==ABORTING
Abort trap: 6
```

**Comment by PhilMiller** [comment]
Here's the GIUH failure seen in `test_unit`:

```
[ RUN      ] Formulation_Manager_Test.basic_reading_1
=================================================================
==4667==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62a0000061f8 at pc 0x558fc5f749f9 bp 0x7ffc942601c0 sp 0x7ffc942601b0
READ of size 8 at 0x62a0000061f8 thread T0
    #0 0x558fc5f749f8 in giuh::giuh_kernel_impl::interpolate_regularized_cdf() /home/runner/work/ngen/ngen/src/core/catchment/giuh/GIUH.cpp:134
    #1 0x558fc591e0f0 in giuh::giuh_kernel_impl::giuh_kernel_impl(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<double, std::allocator<double> >, std::vector<double, std::allocator<double> >, unsigned int) /home/runner/work/ngen/ngen/include/core/catchment/giuh/GIUH.hpp:100
    #2 0x558fc5f7b6b2 in giuh::giuh_kernel_impl::giuh_kernel_impl(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<double, std::allocator<double> >, std::vector<double, std::allocator<double> >) /home/runner/work/ngen/ngen/include/core/catchment/giuh/GIUH.hpp:123
    #3 0x558fc5f7b6b2 in giuh::GiuhJsonReader::build_giuh_kernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) /home/runner/work/ngen/ngen/src/core/catchment/giuh/GiuhJsonReader.cpp:16
    #4 0x558fc5f7d31c in giuh::GiuhJsonReader::get_giuh_kernel_for_id(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/runner/work/ngen/ngen/src/core/catchment/giuh/GiuhJsonReader.cpp:78
    #5 0x558fc5f50369 in realization::Tshirt_Realization::create_formulation(boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, geojson::JSONProperty, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, geojson::JSONProperty> > >*) /home/runner/work/ngen/ngen/src/realizations/catchment/Tshirt_Realization.cpp:311
    #6 0x558fc5a169a2 in realization::Formulation_Manager::construct_formulation_from_tree(simulation_time_params&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, utils::StreamHandler) (/home/runner/work/ngen/ngen/cmake_build/test/test_unit+0x4ef9a2)
    #7 0x558fc5a1d2dc in realization::Formulation_Manager::read(std::shared_ptr<geojson::FeatureCollection>, utils::StreamHandler) (/home/runner/work/ngen/ngen/cmake_build/test/test_unit+0x4f62dc)
    #8 0x558fc597876c in Formulation_Manager_Test_basic_reading_1_Test::TestBody() /home/runner/work/ngen/ngen/test/realizations/Formulation_Manager_Test.cpp:656
    #9 0x558fc5af65df in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2599
    #10 0x558fc5af65df in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2635
    #11 0x558fc5abfffa in testing::Test::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2674
    #12 0x558fc5ac0801 in testing::TestInfo::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2853
    #13 0x558fc5ac1b85 in testing::TestSuite::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:3012
    #14 0x558fc5ad667e in testing::internal::UnitTestImpl::RunAllTests() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:5870
    #15 0x558fc5af7500 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2599
    #16 0x558fc5af7500 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2635
    #17 0x558fc5ac0ad4 in testing::UnitTest::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:5444
    #18 0x558fc5af9979 in RUN_ALL_TESTS() /home/runner/work/ngen/ngen/test/googletest/googletest/include/gtest/gtest.h:2293
    #19 0x558fc5af9979 in main /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest_main.cc:51
    #20 0x7f9e6515bd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #21 0x7f9e6515be3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    #22 0x558fc55dce54 in _start (/home/runner/work/ngen/ngen/cmake_build/test/test_unit+0xb5e54)

0x62a0000061f8 is located 8 bytes to the left of 22256-byte region [0x62a000006200,0x62a00000b8f0)
allocated by thread T0 here:
    #0 0x7f9e65a571c7 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x558fc5f7ad89 in __gnu_cxx::new_allocator<double>::allocate(unsigned long, void const*) /usr/include/c++/11/ext/new_allocator.h:127
    #2 0x558fc5f7ad89 in std::allocator_traits<std::allocator<double> >::allocate(std::allocator<double>&, unsigned long) /usr/include/c++/11/bits/alloc_traits.h:464
    #3 0x558fc5f7ad89 in std::_Vector_base<double, std::allocator<double> >::_M_allocate(unsigned long) /usr/include/c++/11/bits/stl_vector.h:346
    #4 0x558fc5f7ad89 in std::_Vector_base<double, std::allocator<double> >::_M_create_storage(unsigned long) /usr/include/c++/11/bits/stl_vector.h:361
    #5 0x558fc5f7ad89 in std::_Vector_base<double, std::allocator<double> >::_Vector_base(unsigned long, std::allocator<double> const&) /usr/include/c++/11/bits/stl_vector.h:305
    #6 0x558fc5f7ad89 in std::vector<double, std::allocator<double> >::vector(std::vector<double, std::allocator<double> > const&) /usr/include/c++/11/bits/stl_vector.h:555
    #7 0x558fc5f7ad89 in giuh::GiuhJsonReader::build_giuh_kernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) /home/runner/work/ngen/ngen/src/core/catchment/giuh/GiuhJsonReader.cpp:16
    #8 0x558fc5f7d31c in giuh::GiuhJsonReader::get_giuh_kernel_for_id(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/runner/work/ngen/ngen/src/core/catchment/giuh/GiuhJsonReader.cpp:78
    #9 0x558fc5f50369 in realization::Tshirt_Realization::create_formulation(boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, geojson::JSONProperty, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, geojson::JSONProperty> > >*) /home/runner/work/ngen/ngen/src/realizations/catchment/Tshirt_Realization.cpp:311
    #10 0x558fc5a169a2 in realization::Formulation_Manager::construct_formulation_from_tree(simulation_time_params&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, utils::StreamHandler) (/home/runner/work/ngen/ngen/cmake_build/test/test_unit+0x4ef9a2)
    #11 0x558fc5a1d2dc in realization::Formulation_Manager::read(std::shared_ptr<geojson::FeatureCollection>, utils::StreamHandler) (/home/runner/work/ngen/ngen/cmake_build/test/test_unit+0x4f62dc)
    #12 0x558fc597876c in Formulation_Manager_Test_basic_reading_1_Test::TestBody() /home/runner/work/ngen/ngen/test/realizations/Formulation_Manager_Test.cpp:656
    #13 0x558fc5af65df in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2599
    #14 0x558fc5af65df in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2635
    #15 0x558fc5abfffa in testing::Test::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2674
    #16 0x558fc5ac0801 in testing::TestInfo::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2853
    #17 0x558fc5ac1b85 in testing::TestSuite::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:3012
    #18 0x558fc5ad667e in testing::internal::UnitTestImpl::RunAllTests() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:5870
    #19 0x558fc5af7500 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2599
    #20 0x558fc5af7500 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:2635
    #21 0x558fc5ac0ad4 in testing::UnitTest::Run() /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest.cc:5444
    #22 0x558fc5af9979 in RUN_ALL_TESTS() /home/runner/work/ngen/ngen/test/googletest/googletest/include/gtest/gtest.h:2293
    #23 0x558fc5af9979 in main /home/runner/work/ngen/ngen/test/googletest/googletest/src/gtest_main.cc:51
    #24 0x7f9e6515bd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/runner/work/ngen/ngen/src/core/catchment/giuh/GIUH.cpp:134 in giuh::giuh_kernel_impl::interpolate_regularized_cdf()
Shadow bytes around the buggy address:
  0x0c547fff8be0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c547fff8bf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c547fff8c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c547fff8c10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c547fff8c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c547fff8c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa[fa]
  0x0c547fff8c40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c547fff8c50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c547fff8c60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c547fff8c70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c547fff8c80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

**Comment by PhilMiller** [comment]
We worked out that the GeoPackage container overflow error was likely a false positive resulting from having disabled instrumentation of `wkb.cpp` due to the long compilation times. 

https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow#false-positives

I'm going to add a flag to skip those checks for the time being

**Comment by PhilMiller** [comment]
In the interest of delivering value sooner rather than later, I'm going to narrow this PR to just focus on getting AddressSanitizer integrated in CI. My preliminary testing didn't surface any UBSan reports, while it surfaced many ASan reports, some of which re-surfaced in getting it added to CI, so I think ASan is the more important goal for now.

**Comment by PhilMiller** [comment]
Short-term fix for the GIUH error reports while we wait to remove it: I'll add a suppression

File: include/forcing/DataProviderSelectors.hpp [hunk 1]
```diff
@@ -121,6 +121,12 @@ class CatchmentAggrDataSelector
     std::string id_str; //< the catchment to access data for
 };
 
+
+#if !defined(__APPLE__) && !defined(__llvm__) && defined(__GNUC__) && __GNUC__ < 8
+// Enable a workaround for old GCC versions (but not Clang or Apple's fake-GCC that's actually Clang)
+#define NGEN_SELECTOR_CAST
```
Comment (mattw-nws) [inline include/forcing/DataProviderSelectors.hpp:127]: Out of curiosity did this set of changes fix an issue being reported by the ASAN tests or is it just a target of opportunity to make cleaner?

File: src/core/catchment/giuh/GIUH.cpp [hunk 2]
```diff
@@ -100,6 +100,7 @@ void giuh_kernel_impl::set_interpolation_regularity_seconds(unsigned int regular
     }
 }
 
+__attribute__((no_sanitize("address")))
```
Comment (mattw-nws) [inline src/core/catchment/giuh/GIUH.cpp:103]: Per discussion, intent here is to move forward without excising this zombie code.... but this means yes we need to get on with the disposition of this code somehow or another.

**Review by mattw-nws — APPROVED** [review]
Question on the DataProviderSelectors.hpp changes is just a curiosity--I don't see any actual issues. Merging.

File: include/forcing/DataProviderSelectors.hpp (hunk 1 above)
Comment (PhilMiller) [reply include/forcing/DataProviderSelectors.hpp:127]: Target of opportunity - we knew it was wrong, and I was seeing a warning in compiler output over and over as I was pulling this together.

File: src/core/catchment/giuh/GIUH.cpp (hunk 2 above)
Comment (PhilMiller) [reply src/core/catchment/giuh/GIUH.cpp:103]: I opened #587 about this. Initially assigned to you for disposition

**Pull request merged**